src/tokenizer/standard_tokenizer.rs

```diff
@@ -0,0 +1,231 @@
+use super::BoxTokenStream;
+use super::{Token, TokenStream, Tokenizer};
+use std::str::Chars;
+
+/// Tokenize the text by natural division.
+#[derive(Clone)]
+pub struct StandardTokenizer;
+
+pub struct StandardTokenStream<'a> {
+    text: &'a str,
+    chars: Chars<'a>,
+    front_offset: usize,
+    pre_char: Option<char>,
+    token: Token,
+    len: usize,
+}
+
+impl Tokenizer for StandardTokenizer {
+    fn token_stream<'a>(&self, text: &'a str) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(StandardTokenStream {
+            text,
+            chars: text.chars(),
+            pre_char: None,
+            front_offset: 0,
+            len: text.len(),
+            token: Token::default(),
+        })
+    }
+}
+
+impl<'a> StandardTokenStream<'a> {
+    fn next(&mut self) -> (usize, usize) {
+        let mut len = 0;
+        let mut pre_type = 0;
+
+        if let Some(c) = &self.pre_char {
+            pre_type = Self::char_type(*c);
+            len = c.len_utf8();
+            match pre_type {
+                4 => {
+                    self.front_offset += len;
+                    self.pre_char = None;
+                    return (self.front_offset - len, self.front_offset);
+                }
+                _ => {}
+            }
+        }
+
+        while let Some(c) = self.chars.next() {
+            let tp = Self::char_type(c);
+
+            match (pre_type, tp) {
+                (0, 1) => {
+                    self.front_offset += c.len_utf8();
+                }
+                (0, 2) | (0, 3) => {
+                    len = c.len_utf8();
+                    pre_type = tp;
+                }
+                (0, 4) => {
+                    let c_len = c.len_utf8();
+                    self.front_offset += c_len;
+                    self.pre_char = None;
+                    return (self.front_offset - c_len, self.front_offset);
+                }
+                (2, 1) | (3, 1) => {
+                    let start = self.front_offset;
+                    self.front_offset = self.front_offset + len + c.len_utf8();
+                    self.pre_char = None;
+                    return (start, start + len);
+                }
+                (2, 2) | (3, 3) => {
+                    len += c.len_utf8();
+                }
+                _ => {
+                    self.front_offset += len;
+                    self.pre_char = Some(c);
+                    return (self.front_offset - len, self.front_offset);
+                }
+            }
+        }
+
+        let start = self.front_offset;
+        self.front_offset = self.len;
+        self.pre_char = None;
+        return (start, self.len);
+    }
+
+    /// return type for char
+    /// 1: is alphabeti
+    /// 2: is numeric
+    /// 3: is write space
+    /// 4: others
+    pub fn char_type(c: char) -> usize {
+        match c {
+            ' ' => 1,
```

Comment:
that does not seem correct.

There are a lot of other whitespace chars. For instance.
Accentuated letters are broken too here.
katakana / hiraganas should probably not be split too.
etc.